To read the jsonl file using python, First we should upload it to session storage or upload jsonl file after mounting to google drive. We choose first option and uploaded jsonl file into session storage.

Then we can read it using python 'loads' method from json module.

In [ ]:
import json
file_path = '/content/ecfr-title-12.jsonl'

with open(file_path, 'r') as file:
    for line in file:
        data = json.loads(line)
        print(data)

{'text': '', 'meta': {'chapter': ['II', 'II', 'II', 'II', 'II', 'II', 'II', 'II', 'II', 'II', 'II', 'II', 'II', 'II', 'II', 'II', 'II', 'II', 'II', 'II', 'II', 'II', 'II', 'II', 'II', 'II', 'II', 'II', 'II', 'II', 'II', 'II', 'II', 'II', 'II', 'II', 'II', 'II', 'II', 'II', 'II', 'II', 'II', 'II', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'III', 'VI', 'VI', 'VI', 'VI', 'VI', 'VI', 'VI', 'VI', 'VI', 'VI', 'VI', 'VI', 'VI', 'VI', 'VI', 'VI', 'VI', 'VI', 'VI', 'VI', 'VI', 'VI', 'VI', 'VI', 'VI', 'VI', 'VI', 'VI', 'VI', 'VI', 'VI', 'VI', 'VI', 'VI', 'VI', 'VI', 'VI', 'VI', 'VI', 'VII', 'VII', 'VII', 'VII', 'VII', 'VII', 'VII', 'VII', 'VII', 'VII', 'VII', 'VII',

IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.

Current values:
NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
NotebookApp.rate_limit_window=3.0 (secs)




{'text': 'As used in this subpart: , (a) Affiliate means a company that controls, is controlled by, or is under common control with another company. , (b) Bank means a state member bank. , (c) Company means any corporation, partnership, business trust, association or similar organization, or any other trust (unless by its terms the trust must terminate within twenty-five years or not later than twenty-one years and ten months after the death of individuals living on the effective date of the trust). It does not include any corporation the majority of the shares of which are owned by the United States or by any State, or a qualified family partnership, as defined in section 2(o)(10) of the Bank Holding Company Act of 1956, as amended (12 U.S.C. 1841(o)(10)). , (d) Consumer means an individual who purchases, applies to purchase, or is solicited to purchase from you insurance products or annuities primarily for personal, family, or household purposes. , (e) Control of a company has the s

IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.

Current values:
NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
NotebookApp.rate_limit_window=3.0 (secs)



Now the data is loaded and opened. Now we can explore data with methods available in the packages. Let us see the contents of the data by using for loop.

In [ ]:
import json

jsonl_file_path = 'ecfr-title-12.jsonl'

def inspect_jsonl_dataset(file_path, num_samples=5):
    with open(file_path, 'r') as file:
        for i, line in enumerate(file):
            if i >= num_samples:
                break
            record = json.loads(line)
            print(f"Sample Record {i+1}:")
            print(json.dumps(record, indent=4))
            print("=" * 50)

inspect_jsonl_dataset(jsonl_file_path)

Sample Record 1:
{
    "text": "",
    "meta": {
        "chapter": [
            "II",
            "II",
            "II",
            "II",
            "II",
            "II",
            "II",
            "II",
            "II",
            "II",
            "II",
            "II",
            "II",
            "II",
            "II",
            "II",
            "II",
            "II",
            "II",
            "II",
            "II",
            "II",
            "II",
            "II",
            "II",
            "II",
            "II",
            "II",
            "II",
            "II",
            "II",
            "II",
            "II",
            "II",
            "II",
            "II",
            "II",
            "II",
            "II",
            "II",
            "II",
            "II",
            "II",
            "II",
            "III",
            "III",
            "III",
            "III",
            "III",
            "III",
            "III",
     

With above code, we can get to know how the data is arranged and how it organized in the jsonl dataset. We can see that data is proper format.

In [ ]:
import json

def get_meta_details(jsonl_file):
    num_records = 0
    keys_set = set()
    data_types = {}

    with open(jsonl_file, 'r') as file:
        for line in file:
            record = json.loads(line)
            num_records += 1

            for key, value in record.items():
                keys_set.add(key)
                data_type = type(value).__name__
                if key not in data_types:
                    data_types[key] = set()
                data_types[key].add(data_type)

    print("Number of Records:", num_records)
    print("Keys Present:", keys_set)
    print("Data Types:")
    for key, value in data_types.items():
        print(f"- {key}: {value}")

jsonl_file = 'ecfr-title-12.jsonl'

get_meta_details(jsonl_file)


Number of Records: 4666
Keys Present: {'meta', 'text'}
Data Types:
- text: {'str'}
- meta: {'dict'}


The above code give us some meta details like how many records are there in the dataset and what are the datatypes of data present. Meta data is defined data which contains details of another data. All the information is shown in jsonl file and information of jsonl file is stored in meta field.

In [ ]:
import json

jsonl_file_path = 'ecfr-title-12.jsonl'

def extract_author_and_name(file_path):
    authors = set()
    dataset_names = set()

    with open(file_path, 'r') as file:
        for line in file:
            record = json.loads(line)
            if 'author' in record:
                authors.add(record['author'])
            if 'name' in record:
                dataset_names.add(record['name'])

    return authors, dataset_names

authors, dataset_names = extract_author_and_name(jsonl_file_path)

print("Authors:", authors)
print("Dataset Names:", dataset_names)

Authors: set()
Dataset Names: set()


The above code must give more meta data like author and title from dataset but it is not giving in this case. That means there is no author name and title as a key in the dataset. Apart from the author and title names, we got number of records and datatypes of data present in the dataset as meta data. Additionally, by manual checking, the author of dataset is "Administrator of the National Credit Union Administration" and the dataset name is "Code of Federal Rights". The dataset is also having data in different chapters with chapter_titles, sub chapters with subchapter titles, parts with part_titles and different sections with section_titles in meta data.

In [ ]:
import json

jsonl_file_path = 'ecfr-title-12.jsonl'

def analyze_unstructured_data(file_path):
    text_fields = set()

    with open(file_path, 'r') as file:
        for line in file:
            record = json.loads(line)
            for key, value in record.items():
                if isinstance(value, str):
                    text_fields.add(key)

    return text_fields

unstructured_text_fields = analyze_unstructured_data(jsonl_file_path)

print("Identified Text Fields:", unstructured_text_fields)

Identified Text Fields: {'text'}


The unstrtuctured data will be generally present under text, content or description. But in the above dataset, most of the unstructured data is present under text field and it is present as group of lines as a paragraph. Unstructured data in this dataset is more information about the organization and different contents in the website.

In [ ]:
import pandas as pd
import json

def load_jsonl_to_dataframe(file_path):
    with open(file_path, 'r') as file:
        json_list = [json.loads(line) for line in file]
    return pd.DataFrame(json_list)

def check_missing_values(df, fields):
    for field in fields:
        missing_count = df[field].isnull().sum()
        print(f"Missing or null values in '{field}': {missing_count}")

file_path = 'ecfr-title-12.jsonl'

df = load_jsonl_to_dataframe(file_path)

check_missing_values(df, ['meta', 'text'])

Missing or null values in 'meta': 0
Missing or null values in 'text': 0


The above code is for checking for null values in two fields which are in the given jsonl data. The output tells us there are no missing values which make the dataset a complete one.

In [ ]:
import pandas as pd
import json

def load_jsonl_to_dataframe(file_path):
    with open(file_path, 'r') as file:
        json_list = [json.loads(line) for line in file]
    return pd.DataFrame(json_list)

def check_data_consistency(df):
    inconsistent_meta_count = 0
    empty_text_count = 0

    for index, row in df.iterrows():
        if 'title' not in row['meta'] or 'part' not in row['meta']:
            inconsistent_meta_count += 1

        if not row['text'].strip():
            empty_text_count += 1

    print(f"Records with inconsistent 'meta' field: {inconsistent_meta_count}")
    print(f"Records with empty 'text' field: {empty_text_count}")

file_path = 'ecfr-title-12.jsonl'

df = load_jsonl_to_dataframe(file_path)

check_data_consistency(df)

Records with inconsistent 'meta' field: 4666
Records with empty 'text' field: 1


Above python code is for checking if the dataset is having consistency or not. The output says that the meta field contents were inconsistent as for all data entries like 4666 data entries, there was inconsistency and there was an empty text field. We will be evaluating quality of data using factors like 'completeness', 'consistency', 'integrity' and 'accuracy'. In this example, we checked 'completeness' and 'consistency'. If data is following both factors then it will be a consistent data. In this case, the dataset we took is not a quality one as it is having lot of inconsistencies in 'meta' field and there is one missing value in 'text' field.

Below code is for using exploratory method using Scentence Embedding Visualization

In [ ]:
sampled_df = df.sample(n=1000, random_state=42)

texts = sampled_df['text'].tolist()
labels = sampled_df['meta'].apply(lambda x: x.get('label')).tolist()


In [ ]:
# Sample n=1000 rows from the dataset
sampled_df = df.sample(n=1000, random_state=42)  # Use a fixed random state for reproducibility

# Extract texts and labels from the sampled data
texts = sampled_df['text'].tolist()
labels = sampled_df['meta'].apply(lambda x: x.get('label')).tolist()

In [ ]:
from sentence_transformers import SentenceTransformer

model = SentenceTransformer('distilbert-base-nli-stsb-mean-tokens')

def get_embeddings(sentences):
    return model.encode(sentences, batch_size=32, show_progress_bar=True)

embeddings = get_embeddings(texts)

Batches:   0%|          | 0/32 [00:00<?, ?it/s]

In [ ]:
import pandas as pd

# Export embeddings
pd.DataFrame(embeddings).to_csv('output.tsv', sep='\t', index=False, header=None)

# Export metadata
metadata_df = pd.DataFrame({'text': texts, 'label': labels})
metadata_df.to_csv('metadata.tsv', sep='\t', index=False, header=False)
